### PR TITLE
Proxy for datanode rest, jwt auth filter for datanode

### PR DESCRIPTION
**Pull request by todvora**
This PR adds a proxy to the graylog server. The proxy allows accessing datanode rest apis. To secure the apis, jwt auth filter has been introduced. Additionally the nodes service now registers each datanode with its public rest api address, so the proxy knows where to connect. 
/nocl

## Motivation and Context
We need to be able to access the datanode rest apis from gl server. Accessing it via proxy is the easiest and a safe way to do so.

## How Has This Been Tested?
Added unit tests, existing set of unit and integration tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import java.security.Key;
+import java.util.Collections;
+import java.util.Optional;
+
+/**
+ * This is an authorization filter that  first try to verify presence and validity of a JWT token. If there is no
+ * bearer token available, it will fallback to basic auth (or whatever filter is configured as fallback).
+ * Allowing both auth methods allows easy access directly from CLI or browser and machine-machine communication from the graylog server.
+ */
+public class DatanodeAuthFilter implements ContainerRequestFilter {
```
Comment (luk-kaminski) [inline data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java:44]: **We've been talking with Tomas a little bit about it on Slack.
Tomas asked me to put my comments here. It is not a strict request, rather an idea that could be considered.**

1. The thing I'd do is to separate this stuff to 2 separate classes, moving JWT-related stuff to some new class and making `DatanodeAuthFilter` responsible for trying some auth method and using other as a fallback.
But the concrete method - JWT - I'd have in another class, passed to `DatanodeAuthFilter` via constructor.

2. I would have an interface, let's say `AuthVerification` and its implemenation `JwtAuthVerification`. In the test I'd mock `AuthVerification` without referring to any JWT-like details. I'd test `JwtAuthVerification` in a separate unit test.
You have a "classicist" test - like mini-integration test, which is enforced by the different coding style.

3. That's generally another example of slight differences of our coding styles. I try more to keep "single responsibility" from the start (probably I fail it often as well...), you would rather sacrifice this principle at the early stage, just not to have too much code that you deem unnecessary at that point of time.

Comment (todvora) [reply data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java:36]: Thanks @luk-kaminski , I refactored the mentioned parts of the code and both the code and tests are now cleaner and more readable.

File: data-node/src/test/java/org/graylog/datanode/initializers/DatanodeAuthFilterTest.java [hunk 2]
```diff
@@ -40,30 +33,33 @@ class DatanodeAuthFilterTest {
 
     @Test
     void testValidJwtToken() throws IOException {
```
Comment (luk-kaminski) [inline data-node/src/test/java/org/graylog/datanode/initializers/DatanodeAuthFilterTest.java:35]: You can get rid of `Jwt` in the method name now! :)

File: data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+
+    public DatanodeAuthFilter(ContainerRequestFilter fallbackFilter, AuthTokenValidator tokenVerifier) {
+        this.fallbackFilter = fallbackFilter;
+        this.tokenVerifier = tokenVerifier;
+    }
+
+    private Optional<String> getBearerHeader(ContainerRequestContext requestContext) {
+        final MultivaluedMap<String, String> headers = requestContext.getHeaders();
+        return headers.getOrDefault(AUTHORIZATION_PROPERTY, Collections.emptyList())
```
Comment (moesterheld) [inline data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java:52]:
```suggestion
        return headers.getOrDefault(ContainerRequest.AUTHORIZATION, Collections.emptyList())
```

File: data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+/**
+ * This is an authorization filter that  first try to verify presence and validity of a bearer token. If there is no
+ * bearer token available, it will fallback to basic auth (or whatever filter is configured as fallback).
+ * Allowing both auth methods allows easy access directly from CLI or browser and machine-machine communication from the graylog server.
+ */
+public class DatanodeAuthFilter implements ContainerRequestFilter {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DatanodeAuthFilter.class);
+    private static final String AUTHORIZATION_PROPERTY = "Authorization";
```
Comment (moesterheld) [inline data-node/src/main/java/org/graylog/datanode/initializers/DatanodeAuthFilter.java:39]:
```suggestion
```
Can be removed if `ContainerRequest.AUTHORIZATION` is used (see below)

File: data-node/src/main/java/org/graylog/datanode/management/OpensearchProcessService.java [hunk 5]
```diff
@@ -22,6 +22,7 @@
 import org.graylog.datanode.Configuration;
 import org.graylog.datanode.configuration.DatanodeConfiguration;
 import org.graylog.datanode.process.OpensearchConfiguration;
+import org.graylog.datanode.process.StateMachineTracer;
```
[the remark this answers is not in the document]
Comment (todvora) [reply data-node/src/main/java/org/graylog/datanode/management/OpensearchProcessService.java:25]: Thanks, forgotten from experimenting with the structure of the change. Will remove it now.

File: graylog2-server/src/main/java/org/graylog2/rest/resources/datanodes/DatanodeRestApiProxy.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+                .filter(DataNodeDto::isLeader)
+                .map(DataNodeDto::getRestApiAddress)
+                .findFirst()
+                .map(url -> StringUtils.removeEnd(url, "/"))
+                .orElseThrow(() -> new IllegalStateException("No datanode present"));
+
+        final Request builder = new Request.Builder()
+                .url(host + "/" + request.path())
+                .addHeader("Authorization", authTokenProvider.get())
+                .method(request.method(), getBody(request))
```
Comment (moesterheld) [inline graylog2-server/src/main/java/org/graylog2/rest/resources/datanodes/DatanodeRestApiProxy.java:64]: `getBody` returns an empty Body, causing the following exception for GET requests
```
java.lang.IllegalArgumentException: method GET must not have a request body.
	at okhttp3.Request$Builder.method(Request.kt:258) ~[okhttp-4.12.0.jar:?]
	at org.graylog2.rest.resources.datanodes.DatanodeRestApiProxy.request(DatanodeRestApiProxy.java:64) ~[classes/:?]
```

**Review by moesterheld — CHANGES_REQUESTED** [review]
found one error and some minor, but would like to like over it once more later

File: graylog2-server/src/main/java/org/graylog2/rest/resources/datanodes/DatanodeRestApiProxy.java (hunk 6 above)
Comment (todvora) [reply graylog2-server/src/main/java/org/graylog2/rest/resources/datanodes/DatanodeRestApiProxy.java]: Good catch, thank you! Fixed now.

**Review by moesterheld — APPROVED** [review]
works as expected. Thanks for implementing that.

**Pull request merged**